server/playground.rs

```diff
@@ -0,0 +1,139 @@
+use hubcaps::gists;
+use iron::{status, IronResult, Request, Response};
+use router::Router;
+use serde::{Deserialize, Serialize};
+use serde_json;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::io::Read;
+use tokio::runtime::Runtime;
+
+use crate::errors::Error;
+use crate::json;
+use crate::middleware::GistsMiddleware;
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Playground {
+    id: Option<String>,
+    messages: String,
+    variables: serde_json::Value,
+    setup: serde_json::Value,
+}
+
+pub fn get(req: &mut Request) -> IronResult<Response> {
+    let gists_middleware = req.extensions.get::<GistsMiddleware>().unwrap();
+    let gists = &gists_middleware.gists;
+    let params = req.extensions.get::<Router>().unwrap();
+    let id = params.find("id").expect("No route parameter called id");
+    let mut rt = match Runtime::new() {
+        Ok(rt) => rt,
+        Err(_) => return json::error(status::ServiceUnavailable, Error::Runtime),
+    };
+    let gist = match rt.block_on(gists.get(id)) {
+        Ok(gist) => gist,
+        Err(hubcaps::errors::Error(hubcaps::errors::ErrorKind::Fault { code, .. }, _))
+            if code == 404 =>
+        {
+            return json::error(status::NotFound, Error::NotFound)
+        }
+        Err(_) => return json::error(status::InternalServerError, Error::Fetching),
+    };
+    match Playground::try_from(gist) {
+        Ok(playground) => json::respond(playground),
+        Err(err) => json::error(status::InternalServerError, err),
+    }
+}
+
+pub fn create(req: &mut Request) -> IronResult<Response> {
+    let gists_middleware = req.extensions.get::<GistsMiddleware>().unwrap();
+    let gists = &gists_middleware.gists;
+    let mut payload = String::new();
+    if let Err(_) = req.body.read_to_string(&mut payload) {
+        return json::error(status::InternalServerError, Error::ReadingRequest);
+    }
+    let playground = match serde_json::from_str::<Playground>(&payload) {
+        Ok(playground) => playground,
+        Err(_) => return json::error(status::InternalServerError, Error::Deserializing),
+    };
+    let options = match gists::GistOptions::try_from(playground) {
+        Ok(options) => options,
+        Err(err) => return json::error(status::InternalServerError, err),
+    };
+    let mut rt = match Runtime::new() {
+        Ok(rt) => rt,
+        Err(_) => return json::error(status::ServiceUnavailable, Error::Runtime),
+    };
```

Comment:
`unwrap_or_else`

Reply:
Same as in the other comment, I'd like to benefit from early returns in this code and I don't think `unwrap_or_else` would allow me to do it. Is there some other `Result` combinator which would be a better choice? I understand `and_then` can be used to chain all functions returning the same `Result` type and then handle the error at the end of the chain.